### PR TITLE
Add game logic for Board cell clicks

### PR DESCRIPTION
**Pull request by davidl21**
**Summary** 

Implements core game interaction logic upon user clicking a Board cell 

- adds game state validation to prevent out of bounds moves
- sets appropriate cell flags on click
- reveals single or multiple cells according to Minesweeper rules

**Testing**

- adds new unit test
- tests out of bounds edge case
- tests case where a user might click on a revealed cell multiple times
- run tests with `python -m pytest`

**To Do**

- add logic for when player loses the game, revealing all squares
- add error classes for better HTTP responses

### CODE REVIEW

File: backend/models.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import random
+from typing import Optional
 
 class Cell:
     def __init__(self, x: int, y: int):
```
Comment (cxindyli) [inline backend/models.py:5]: extract Cell class into different types of cells so that mine and regular-cell functions and attributes can be separated

File: backend/models.py [hunk 2]
```diff
@@ -27,6 +28,7 @@ def __init__(self, cols: int, rows: int, num_mines: int):
         self.rows = rows
         self.num_mines = num_mines
         self.grid = [[Cell(r, c) for c in range(self.cols)] for r in range(self.rows)]
+        self.offsets = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, -1), (-1, 1)]
```
Comment (cxindyli) [inline backend/models.py]: don't have to store this on every `Board `instance; instead you can use a class variable

File: backend/models.py [hunk 3]
```diff
@@ -27,6 +28,7 @@ def __init__(self, cols: int, rows: int, num_mines: int):
...
 
     def place_mines(self):
         all_positions = []
```
Comment (cxindyli) [inline backend/models.py]: for lines 34-40, can extract into another function

Comment (cxindyli) [reply backend/models.py]: also, i don't think it's good to initialize a new instance variable within a random function in a class. you should probably have all instance variables defined clearly in init, so here you could probably just set self.mine_positions to an empty sequence in your init method and then set mine_positions when you get the actual mine positions

File: backend/models.py [hunk 4]
```diff
@@ -44,9 +46,7 @@ def place_mines(self):
             self.compute_neighbors(r, c)
 
     def compute_neighbors(self, r: int, c:int) -> None:
```
Comment (cxindyli) [inline backend/models.py]: probably rename this to something else; compute_neighbors gives off the impression that you're simply computing something but you're actually setting some important variables here and making changes to the object

File: backend/models.py [hunk 5]
```diff
@@ -44,9 +46,7 @@ def place_mines(self):
             self.compute_neighbors(r, c)
```
Comment (cxindyli) [inline backend/models.py]: i think this should go inside MineCell functionality, where you can just call MineCell.incrementNeighborMineCount

File: backend/models.py (hunk 4 above)
Comment (cxindyli) [reply backend/models.py]: probably rename to `incrementNeighborMineCounts`?

File: backend/models.py (hunk 5 above)
Comment (cxindyli) [reply backend/models.py]: so instead the code would look like:
```suggestion
            for r, c in mine_positions:
              self.grid[r][c] = new MineCell(lalala)
              mine.incrementNeighborMineCount
```

File: backend/models.py [hunk 6]
```diff
@@ -27,6 +28,7 @@ def __init__(self, cols: int, rows: int, num_mines: int):
         self.rows = rows
```
Comment (cxindyli) [inline backend/models.py:30]: also can't comment on the above code but perhaps rename `add_neighbor `to `increment_neighbor_mine_count`

File: backend/models.py [hunk 7]
```diff
@@ -44,9 +46,7 @@ def place_mines(self):
             self.compute_neighbors(r, c)
 
     def compute_neighbors(self, r: int, c:int) -> None:
-        offsets = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, -1), (-1, 1)]
-
-        for r_offset, c_offset in offsets:
+        for r_offset, c_offset in self.offsets:
             new_r, new_c = r + r_offset, c + c_offset
 
             if self.is_inbound(r, c):
```
Comment (cxindyli) [inline backend/models.py:52]: don't just return None, should return an Error class that has more specific information

File: backend/models.py [hunk 8]
```diff
@@ -62,30 +62,52 @@ def handle_click(self, r: int, c:int) -> list[Cell]:
...
+                revealed_cells = self.reveal(r, c) 
 
-            return 
+            return revealed_cells 
         else:
-            # throw out of bounds error
-            return
+            return []
+
+    def reveal(self, r: int, c: int) -> list[Cell]:
```
Comment (cxindyli) [inline backend/models.py]: based on this function name it's hard to deduce what it actually does / returns

File: backend/models.py [hunk 9]
```diff
@@ -62,30 +62,52 @@ def handle_click(self, r: int, c:int) -> list[Cell]:
         revealed_cells = []
 
         if self.is_inbound(r, c):
```
Comment (cxindyli) [inline backend/models.py]: how is it possible for someone to click out of bounds and still give you an (r, c)?

File: backend/models.py [hunk 10]
```diff
@@ -62,30 +62,52 @@ def handle_click(self, r: int, c:int) -> list[Cell]:
...
-                # do nothing, handle later
-                return
+                return []
         
             if cell.is_flagged:
-                # remove flag
-                return
+                cell.toggle_flag()
 
             if cell.is_mine:
```
Comment (cxindyli) [inline backend/models.py]: here you'd check for the type of cell instead of just boolean like so:

```
cell match {
case MineCell => Result(Result.GameOver)
case RegularCell => self.reveal 
                                 Result(Result.Continue)
}
```
store list[revealed] as an attribute for the board, and then change as it's changed

File: backend/models.py (hunk 1 above)
Comment (davidl21) [reply backend/models.py:5]: I'm thinking of leaving the Cell class as one class because it seems like Mine and Non-Mine cells have almost no behavioral differences and it might make typechecking more difficult, and only difference is their behavior when the game ends (when clicking on a mine)

File: backend/models.py (hunk 2 above)
Comment (davidl21) [reply backend/models.py]: done

File: backend/models.py (hunk 3 above)
Comment (davidl21) [reply backend/models.py]: since the storing of mine_positions is only useful for the function of placing mines, would it make sense to have it only for the function and not as an instance variable because we don't need it to persist through the game?

maybe for game victory, I'll store the count of revealed cells and the total number of non-mine cells, and check if they're equal to each other?

File: backend/models.py (hunk 4 above)
Comment (davidl21) [reply backend/models.py]: done

File: backend/models.py (hunk 5 above)
Comment (davidl21) [reply backend/models.py]: I think a potential issue is that minecells don't have a view of the entire board, so incrementingneighbormine count using r, c indices doesn't work unless I can access the Board's self.grid[][] or something similar

File: backend/models.py (hunk 6 above)
Comment (davidl21) [reply backend/models.py:30]: done

File: backend/models.py (hunk 8 above)
Comment (davidl21) [reply backend/models.py]: fixed, changed to get_revealed_cells

File: backend/models.py (hunk 9 above)
Comment (davidl21) [reply backend/models.py]: u right...
